benchmark/benchmark.py

```diff
@@ -0,0 +1,117 @@
+"""Benchmarking utility for PennyLane.
+"""
+# pylint: disable=import-outside-toplevel,invalid-name
+import functools
+import importlib
+import subprocess
+import sys
+
+import numpy as np
+
+import pennylane as qml
+
+
+def timing(func, *, number=10, repeat=5):
+    """Time the given function.
+
+    Args:
+        func (callable[None, Any]): function to time
+        number (int): number of loops per test
+        repeat (int): the number the timing test is run
+    """
+    import timeit
+
+    print('{} loops, {} runs'.format(number, repeat))
+    res = timeit.repeat(func, number=number, repeat=repeat, globals=globals())
+    print("Timing per loop:", np.array(res) / number)
+
+
+def plot(title, kernels, n_range):
+    """Plot timings as a function of the parameter n.
+
+    Args:
+        title (str): plot title
+        kernels (Sequence[callable[Any, Any]]): parametrized benchmarks to time
+        n_range (Sequence[Any]): values the benchmark parameter n takes
+    """
+    import perfplot
+
+    labels = [k.__doc__ for k in kernels]
+    perfplot.show(
+        setup=lambda n: n,
+        kernels=kernels,
+        labels=labels,
+        n_range=n_range,
+        xlabel="n",
+        title=title,
+        # logx="auto",  # set to True or False to force scaling
+        # logy="auto",
+        # equality_check=numpy.allclose,  # set to None to disable "correctness" assertion
+        # automatic_order=True,
+        # colors=None,
+        # target_time_per_measurement=1.0,
+        # time_unit="s",  # set to one of ("auto", "s", "ms", "us", or "ns") to force plot units
+        # relative_to=1,  # plot the timings relative to one of the measurements
+        # flops=lambda n: 3*n,  # FLOPS plots
+    )
+
+
+def profile(func, identifier):
+    """Profile the given function.
+
+    Args:
+        func (callable[None, Any]): function to profile
+        identifier (str): identifying part of the name of the file containing the results
+    """
+    import cProfile
+    import pstats
+
+    pr = cProfile.Profile()
+    pr.enable()
+
+    for _ in range(20):
```

Comment:
Anything special about 20? Can this also be set by an optional kwarg in `profile`?

Reply:
Sure, I'll change it. The idea is to repeat the benchmark a sufficient number of times to get rid of stochasticity in the timings. Or maybe we could use a timer and keep repeating the benchmark until at least, say, 10 seconds has passed?

Reply:
probably the first option (fixed number of repetitions), as the second method might lead to a different number of benchmarks being executed within different 10 second windows

Since 20 is a magic number, can we define it at the top of the file in all caps?

Reply:
When doing profiling we're mostly interested in the fraction of time spent in various functions, trying to locate bottlenecks, so the number of repeats isn't that important. For raw performance benchmarking I think the timing functionality is more useful.